editoast: database connection pool tools

Wadjetz opened the pull request:
part of #6980

- [x] Creation of custom ConnectionPool
- [x] Isolation of tests with test_connection transaction (that do not commit the transaction)
- [x] Exemple of usage in documents api
- [x] Test App Builder

leovalais commented on editoast/src/connection_pool.rs:
It'd be nice to define `type` aliases for diesel objects (the Pool and the Object connection).

leovalais commented on editoast/src/modelsv2/mod.rs:
```diff
@@ -29,6 +29,7 @@ pub use rolling_stock_image::RollingStockSeparatedImageModel;
 pub use rolling_stock_model::RollingStockModel;
 pub use study::Study;
 
+#[allow(unused_imports)]
```
You can remove unused imports.

leovalais commented on editoast/src/views/documents.rs:
```diff
@@ -49,9 +55,10 @@ pub enum DocumentErrors {
-async fn get(db_pool: Data<DbPool>, document_key: Path<i64>) -> Result<HttpResponse> {
+async fn get(db_pool: Data<ConnectionPool>, document_key: Path<i64>) -> Result<HttpResponse> {
+    let mut conn = db_pool.get().await?;
+    let conn = &mut conn.get().await?;
```
I'd really like to avoid breaking the API outside of tests and low-level models calls.

leovalais commented on editoast/src/views/mod.rs:
```diff
@@ -475,6 +477,40 @@ mod tests {
         create_test_service_with_core_client(CoreClient::default()).await
     }
 
+    pub async fn create_test_service_with_connection_pool(
```
It's probably a temporary function while all tests aren't updated, but the goal is to change `create_test_service`

leovalais commented on editoast/src/views/mod.rs:
```diff
@@ -475,6 +477,40 @@ mod tests {
+            .url()
+            .expect("cannot get postgres config url");
+        let config =
+            AsyncDieselConnectionManager::<diesel_async::AsyncPgConnection>::new(pg_config_url);
```
This needs to be made opaque as well, we don't want any async dependency outside of models.

leovalais commented on editoast/src/views/documents.rs:
```diff
@@ -131,72 +131,74 @@ mod tests {
-    async fn document_delete(#[future] document_example: TestFixture<Document>) {
-        let document_example = document_example.await;
-        let service = create_test_service().await;
+    async fn document_delete() {
```
Oh, when I wrote that we want to get rid of the fixtures in the issue, I meant two things:

1. Remove the TestFixture<> type as its Drop behavior is not necessary after you changes
2. Get rid of the `#[fixture]` and `#[future]` annotations.

But we still want functions that insert test data in the database. Tests are supposed to be isolated: calling an endpoint to test another seems fragile to me. It'd be great if the test could take to the following steps:

```rust
// 1. Have a test document inserted in the db
let document: Document = test_document().await.unwrap();

// 2. (opt) maybe get a ConnectionPool (unique by design from the Arc)
let pool = test_pool().await!;

// 3. test the endpoint
DELETE /doc

// 4. test the deletion
assert!(!document.exists(pool.get().await!).await!);


// not 5. no cleanup to do since we're in a test transaction
```

leovalais commented on editoast/src/connection_pool.rs:
I believe that the `#[cfg(test)]` and `#[cfg(not(test))]` should have the same API (no `'a` for one and not for the other, etc.) otherwise I'm sure we'll run into problems later.

leovalais reviewed (commented):
Thanks for starting on that issue it's really not an easy task.
I have a few comments, mostly API-related.

woshilapin commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,85 @@
+    pub fn new(pool: Pool<AsyncPgConnection>) -> Self {
+        Self {
+            pool: Arc::new(pool),
+            #[cfg(all(test))]
```
```suggestion
            #[cfg(test)]
```

woshilapin commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    #[cfg(test)]
+    pub async fn new_for_tests(
```
Why do we need a `new` and a `new_for_tests`? Why not only a `new` feature-gated? I see that you need `async` and `Result` additionally, but maybe `Result` can be avoided (unwrapping is probably ok for tests). And for the `async` part, I believe we should be able to run the nested functions with `Handle::current().block_on(whatever_async_function())`.

woshilapin commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    #[cfg(not(test))]
```
Maybe to avoid so many feature-gated functions, you could do only 3 parts like the following.

```rust
impl DbConnectionPoolV2 {
    pub fn new() -> Self {}
}

#[cfg(not(test))]
impl DbConnectionPoolV2 {
    pub async fn get() {}
}

#[cfg(test)]
impl DbConnectionPoolV2 {
    pub async fn new_for_tests() {}
    pub async fn get() {}
}
```

Note that it might become less important if some of the functions disappear.

woshilapin commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,85 @@
+}
+
+impl DbConnectionPoolV2 {
+    pub fn new(pool: Pool<AsyncPgConnection>) -> Self {
```
Wouldn't it make sense that the `new` function takes a URL instead of a Pool? Because it looks weird that for building a pool (`DbConnectionPoolV2`), you need a pool (`Pool<AsyncPgConnection>`). It seems that the API is maybe not at the right level.

woshilapin commented on editoast/src/views/mod.rs:
```diff
@@ -437,6 +439,39 @@ mod tests {
         }};
     }
 
+    pub async fn create_test_connection_pool_v2() -> DbConnectionPoolV2 {
+        let pg_config_url = PostgresConfig::default()
+            .url()
+            .expect("cannot get postgres config url");
+        create_connection_pool_for_tests_v2(pg_config_url)
+            .await
+            .expect("cannot create test pool")
+    }
+
+    pub async fn create_test_service_with_connection_pool_v2(
+        connection_pool: DbConnectionPoolV2,
+    ) -> impl Service<Request, Response = ServiceResponse<BoxBody>, Error = Error> {
+        let redis = RedisClient::new(RedisConfig::default()).expect("cannot get redis client");
+
+        // Custom Json extractor configuration
+        let json_cfg = JsonConfig::default()
+            .limit(250 * 1024 * 1024) // 250MB
+            .error_handler(|err, _| InternalError::from(err).into());
+
+        let core: CoreClient = CoreClient::default();
+        let app = App::new()
+            .wrap(NormalizePath::trim())
+            .app_data(json_cfg)
+            .app_data(Data::new(connection_pool))
+            .app_data(Data::new(redis))
+            .app_data(Data::new(CHashMap::<i64, InfraCache>::default()))
+            .app_data(Data::new(MapLayers::parse()))
+            .app_data(Data::new(MapLayersConfig::default()))
+            .app_data(Data::new(core))
+            .service(routes());
+        init_service(app).await
+    }
+
```
Shouldn't these 2 functions be protected by a `#[cfg(test)]`?

woshilapin replied:
Already in a `tests` module 🤦 I'm resolving.

woshilapin reviewed (commented):
It'd be nice to have all of this new setup. 

I've made a few comments because I believe we might not need so many new functions, specific for tests... Let's discuss it, there might be some constraints that I'm not aware of.

woshilapin reviewed (approved):
Nice clean up, looks good to me.

Just to know, a pool size of 1 for the tests was not possible? I'm just curious if you tried it as to why it doesn't work.

Wadjetz committed:
editoast: refactor connection pool for tests

hamz2a reviewed (approved):
Good job, well done!

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_error.rs:
```diff
@@ -0,0 +1,29 @@
+    fn get_type(&self) -> &str {
+        "editoast:ConnectionPoolError"
+    }
+}
```
Move this block into `src/errors.rs`

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+/// ## Insert the required data before testing the API service:
+///     let document = create_document(&pool).await;
+/// ## Call the API service
+/// ## Test the API service response
```
You can use a numbered list and code markers ;)

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+/// ## Test the API service response
+///
+/// The connection pool in test mode will not commit any changes to the database,
+/// all data will be deleted after the test
```
The fact that we `get` the connection instantly at the pool creation for `test` needs to be documented, in order to avoid any inner `db_pool.get()` in models' tests.

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+/// all data will be deleted after the test
+#[derive(Clone)]
+pub struct DbConnectionPoolV2 {
+    pub pool: Arc<Pool<AsyncPgConnection>>,
```
```suggestion
    pub(in crate::modelsv2) pool: Arc<Pool<AsyncPgConnection>>,
```

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_error.rs:
```diff
@@ -0,0 +1,29 @@
+#[derive(Debug, Error)]
+pub enum DbConnectionError {
+    #[error(transparent)]
+    DeadpoolPool(diesel_async::pooled_connection::deadpool::PoolError),
```
```suggestion
    DeadpoolPool(#[from] diesel_async::pooled_connection::deadpool::PoolError),
```

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+impl DbConnectionPoolV2 {
+    pub async fn create(pool: Pool<AsyncPgConnection>) -> Result<Self, DbConnectionError> {
+        use diesel_async::AsyncConnection;
+        let mut conn = pool.get().await.map_err(DbConnectionError::DeadpoolPool)?;
```
```suggestion
        let mut conn = pool.get().await?;
```

With `#[from]` annotations in the error enum this may be done implicitly.

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+
+#[cfg(test)]
+impl DbConnectionPoolV2 {
+    pub async fn create(pool: Pool<AsyncPgConnection>) -> Result<Self, DbConnectionError> {
```
```suggestion
    pub async fn try_from_pool(pool: Pool<AsyncPgConnection>) -> Result<Self, DbConnectionError> {
```

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+        })
+    }
+
+    pub async fn create_from_url(url: Url, pool_size: usize) -> Result<Self, DbConnectionError> {
```
```suggestion
    pub async fn try_initialize(url: Url) -> Result<Self, DbConnectionError> {
```

We can enforce a single connection for the poolv2 as it won't be shared between unit tests.

leovalais commented on editoast/src/modelsv2/connection_pool.rs:
```diff
@@ -23,6 +29,14 @@ pub fn create_connection_pool(url: Url, max_size: usize) -> DbConnectionPool {
 }
 
+#[cfg(test)]
+pub fn create_connection_pool_for_tests(url: Url) -> DbConnectionPool {
```
```suggestion
pub fn create_shared_connection_pool_for_tests(url: Url) -> DbConnectionPool {
```

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+            .map_err(DbConnectionError::DeadpoolBuildPool)?;
+        Self::create(pool).await
+    }
+
```
```suggestion
pub fn for_test() -> Self { ... }
```

We need a function with a default test URL since we'll need it in crates where `client::PostgresConfig` is not defined. You could also read a env variable `OSRD_TEST_PG_URL` in case someone has a weird setup.


You can `block_on` and `expect` on this one since we're in a test anyway.

leovalais commented on editoast/src/views/mod.rs:
```diff
@@ -442,6 +443,41 @@ mod tests {
+            .limit(250 * 1024 * 1024) // 250MB
+            .error_handler(|err, _| InternalError::from(err).into());
+
+        let core: CoreClient = CoreClient::default();
```
At some point we'll need to provide a MockingClient as well. We might as well tackle that now in order to avoid changing the API later on.

We could have the following PAI:

```rust
let db_pool = DbConnectionPoolV2::for_test();
let app = test_app(db_pool.clone(), Default::default());

// or, if we need a MockingClient but don't care about the pool

let core = MockingClient::new();
let app = test_app(Default::default(), core); // where default() = for_test()
```

And this function doesn't need to be `async`, you can `block_on` here.

leovalais replied:
As discussed, maybe a builder would be a better idea. I suggest the following api, feel free to improve it:

```rust
// unused pool, default core client
let app = TestApp::default();

// explicitly specified pool and mocking client
let app = TestApp::new().pool(pool.clone()).core_client(my_mocking_client).build();

// implicitly created pool, default core client
let app = TestApp::default();
let pool: Arc<DbConnectionPool> = app.get_pool(); // When building the app, we could store a ref to the pool, so that we can access it later :)
```

leovalais commented on editoast/src/views/documents.rs:
```diff
@@ -131,72 +134,91 @@ mod tests {
+        let create_response: PostDocumentResponse =
+            call_and_read_body_json(&service, request).await;
+
+        assert!(create_response.document_key > 0);
```
You should try to retrieve the document back with the ID and check that the content matches.

leovalais commented on editoast/src/views/documents.rs:
```diff
@@ -131,72 +134,91 @@ mod tests {
+            .to_request();
+        let response = call_service(&service, request).await;
+
+        assert!(response.status().is_success());
```
You should check the response content.

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,92 @@
+            Err(DbConnectionError::TestConnection)
+        }
+    }
+}
```
A QOL function definition could be

```rust
pub fn get_ok(&self) -> OwnedRwLockWriteGuard<Object<AsyncPgConnection>>
```

with a `blcok_on` and an `expect `

leovalais commented on editoast/src/views/documents.rs:
```diff
@@ -131,72 +134,91 @@ mod tests {
+
+        // In test mode, the connection pool will not commit any changes to the database.
+        // Therefore, it is necessary to create all required data before testing the API service.
+        let document = create_document(&pool).await;
```
When the test setup is small, I'd rather having it inlined in the test. I get why we factorize the infra setup, or train schedules, etc.. but documents are small enough to not suffer from redundancy. Nothing blocking though :)

Wadjetz replied:
I tried to inline the creation of the document, but this caused a deadlock. In the function create_document, we get the connection lock guard, and at the end of the function, the guard gets released. However, if we inline,  the test will holds the guard, and when we try to call the API, it will also try to get the guard and become blocked.

Wadjetz replied:
To release the lock we can also do, but maybe a dedicated function is better ?
```rust
let document = {
    let mut conn = pool.get().await.expect("Failed to get connection pool");
    Document::changeset()
        .data(b"Document post test data".to_vec())
        .content_type(String::from("text/plain"))
        .create(&mut conn)
        .await
        .expect("Failed to create document")
};
```

leovalais replied:
Personally I don't mind the scoping block, as it's a reminder that the connection should be dropped as soon as it's not used anymore.

Alternatively this should work as well:

```rust
let document = Document::changeset()
        .data(b"Document post test data".to_vec())
        .content_type(String::from("text/plain"))
        .create(pool.get_ok().deref_mut())
        .await
        .expect("Failed to create document")
```

Maybe @woshilapin can weigh on this?

woshilapin replied:
Either way is fine with me. The scoping is usually a sign of a guard (but maybe it's only me who sees this as a sign?). Maybe @leovalais solution is more straightforward for those less familiar with the concept of guards?

leovalais reviewed (changes requested):
Nice tricks on the V2 wrapper and on ditching the Connection wrapper for now! I have a few comments on code organization and API names and QOL. 

I spotted a few things I wish were named differently as I found a bit difficult to tell apart what's new, what was already there given the redundancy of the setup phases at both pool and app level. A tad more documentation would be of help as well :)

I'd also rename the module `connection_pool` `modelsv2::connections` as we may also need to customize the future wrapper around the AsyncPgConnection (to foward transactions for example 👀).

Thanks! :)

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,115 @@
+#[cfg(test)]
+impl DbConnectionPoolV2 {
+    /// Get inner pool for retro compatibility
+    pub fn pool(&self) -> Arc<Pool<AsyncPgConnection>> {
```
```suggestion
    pub fn pool_v1(&self) -> Arc<Pool<AsyncPgConnection>> {
```

this way we're sure this function is not here to stay 😆

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
````diff
@@ -0,0 +1,115 @@
+    ///        .await
+    ///        .expect("Failed to create document")
+    /// };
+    /// ```
````
I think rust-analyzer could have trouble on conflicting doctorings. Maybe just comment on the `not(test)` `get` and make a section for the behavior for tests.

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,115 @@
+    pub fn get_ok(&self) -> OwnedRwLockWriteGuard<Object<AsyncPgConnection>> {
+        use tokio::runtime::Handle;
+        Handle::current()
+            .block_on(self.get())
```
```suggestion
        futures::executor::block_on(self.get())
```

leovalais commented on editoast/src/modelsv2/connection_pool/db_connection_pool.rs:
```diff
@@ -0,0 +1,115 @@
+        })
+    }
+
+    pub async fn try_initialize(url: Url, pool_size: usize) -> Result<Self, DbConnectionError> {
```
```suggestion
    pub async fn try_initialize(url: Url) -> Result<Self, DbConnectionError> {
```

as discussed orally

leovalais commented on editoast/src/views/documents.rs:
```diff
@@ -124,80 +127,102 @@ async fn delete(db_pool: Data<DbConnectionPool>, document_key: Path<i64>) -> Res
+            .uri(&format!("/documents/{}", create_response.document_key))
             .to_request();
-        let response: PostDocumentResponse = call_and_read_body_json(&service, request).await;
+        let response = call_and_read_body(&service, request).await;
```
Better use `retrieve` directly

leovalais commented on editoast/src/views/documents.rs:
```diff
@@ -124,80 +127,102 @@ async fn delete(db_pool: Data<DbConnectionPool>, document_key: Path<i64>) -> Res
+            .to_request();
+        let response = call_service(&service, request).await;
+
+        assert_eq!(response.status(), StatusCode::NOT_FOUND);
```
Better use `exists` directly

leovalais reviewed (commented):
We discussed these points orally, just leaving them here for traceability.

leovalais committed:
editoast: add `for_test` function to DbConnectionPoolV2

leovalais committed:
editoast: add DbConnectionPoolV2 documentation and guidelines

leovalais committed:
editoast: add TestAppBuilder for DbConnectionPoolV2 and refactor setups

leovalais committed:
editoast: rename module connection_pool to database

leovalais reviewed (approved):
Approving in order not to block this PR, but since I'm participating now I'd like another review by @woshilapin plz 🙏

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
```diff
@@ -0,0 +1,275 @@
+
+    /// Create a connection pool for testing purposes
+    ///
+    /// You can set the `OSRD_TEST_PG_URL` environment variable to use a custom database url.
```
Did any of you tried to run tests with a Docker setup? Because `localhost` might not work, so we might add `OSRD_TEST_PG_URL` as an environment variable in `docker-compose.yml`.

woshilapin replied:
Tested on Mac, it works!

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
````diff
@@ -0,0 +1,275 @@
+    /// // you may acquire a new connection afterwards
+    /// ```
+    pub async fn get(&self) -> Result<DbConnectionV2, DbConnectionError> {
+        #[cfg(test)]
+        let conn = self.get_test().await;
+        #[cfg(not(test))]
+        let conn = self.get_proxy().await;
````
Why not naming `get_test` and `get_proxy` with the same name (`get_connection`?) to avoid this `#[cfg()]` branch mechanism.

```suggestion
        let conn = self.get_proxy().await;
```

Wadjetz replied:
I think this is for the documentation. We are not sure which version of the doc editor will be chosen, the test or non-test version. By having only one function, this is not a problem.

woshilapin replied:
I do not understand, `get_proxy` and `get_test` don't have a documentation.

Wadjetz replied:
Yes, because they are private functions. If we use cfg(test) or not(test) with two different functions having the same name, which version of the documentation will be shown in the editor? I guess it depends on the configuration of each editor. With only one function without any cfg and proper documentation, we can be sure there are no problems with that.

leovalais replied:
The documentation covers the behavior of the function with in both `test` and `not(test)`. This is intentional since the `not(test)` usage of the connection is limited by how testing works (e.g. no `join!` on connection acquisition).

The problem is with `rust-analyzer` that has `test` on by default, so when we hover a function, its `test` documentation will be displayed. I felt it was weird to document only the `#[cfg(test)] fn get` but I also didn't want to duplicate the whole documentation.

So the `get` function serves as a single entry point, to have the documentation in a single place, and then the branching occurs.

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
```diff
@@ -0,0 +1,275 @@
+        } else {
+            Err(DbConnectionError::TestConnection)
+        }
+    }
```
You already have an entire section of `#[cfg(test)] impl DbConnectionPoolV2 {}` above (lines 50), why not putting this function up there directly?

leovalais replied:
https://github.com/OpenRailAssociation/osrd/pull/7281#discussion_r1611658041

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    #[cfg(not(test))]
+    async fn get_proxy(&self) -> Result<DbConnectionV2, DbConnectionError> {
+        let connection = self.pool.get().await?;
+        Ok(connection)
+    }
```
You already have an entire section of `#[cfg(not(test))] impl DbConnectionPoolV2 {}` above (lines 88), why not putting this function up there directly?

Wadjetz replied:
This is related to the documentation in test or not test mode

woshilapin replied:
I think I do not understand since there is no documentation on this function.

leovalais replied:
https://github.com/OpenRailAssociation/osrd/pull/7281#discussion_r1611658041

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
```diff
@@ -0,0 +1,275 @@
+    ///
+    /// - If you need to open several connections, then the connection must be
+    ///   acquired just before its usage, and dropped just after, **all in the same future**.
+    ///   And these fututres must all be awaited before attempting to acquire a new connection.
```
```suggestion
    ///   And these futures must all be awaited before attempting to acquire a new connection.
```

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
```diff
@@ -0,0 +1,275 @@
+
+    /// Returns an infinite iterator of futures resolving to connections acquired from the pool
+    ///
+    /// Meant to be used in conjunction with `zip` in order instanciate a bunch of tasks to spawn.
```
```suggestion
    /// Meant to be used in conjunction with `zip` in order to instantiate a bunch of tasks to spawn.
```

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
````diff
@@ -0,0 +1,275 @@
+    /// let results = futures::future::try_join_all(operations).await?;
+    /// // you may acquire a new connection afterwards
+    /// ```
+    #[allow(unused)] // TEMPORARY
````
Maybe say why it's temporary so if we ever forget to remove it, the future developer who's discovering this know if it can be removed or not.

```suggestion
    #[allow(unused)] // TEMPORARY until ModelV2 is used
```

woshilapin commented on editoast/src/modelsv2/database/connection_pool.rs:
With the quantity of documentation, and the tricky behaviors we need to be careful about, I'm worried we bring a complexity into production, only for testability purpose. I do not know how to act about that though. Maybe we can discuss it?

woshilapin replied:
Some summary of the discussion.
The limitations given by the connection provider should impact only the tests, not the production environment. One important thing is that : before, endpoints were taking a pool as an input, and were getting a new connection for each request to the DB. The new version will directly give a connection as an input to the endpoints, so the same connection is used for an entire duration of an endpoint handling.

A nice outcome of that new architecture is that if an endpoint tries to do a lot of DB requests, it will not starve the connection pool for other endpoints, meaning the entire application should not be impacted by a greedy endpoint. The drawback might be that endpoints that do need to do a lot of DB requests, might be slower now. It's a different balance, but probably a good one until proven otherwise.

leovalais replied:
Thanks for summarizing our discussion :)

While we agree on the final goal of one connection per endpoint should be the goal, we are still able to retrieve the pool from the endpoints (via the `Data<>` extractor). This is intentional for two reasons:

1. the progressive migration to the DB pool v2, we don't necessarily want to rewrite endpoints that `get` multiple connections _now_
2. maybe there are cases when starvation is deemed acceptable (I'm not sure I see any, but we'll never have a lot of concurrent users, so it may be an easy way to boost an endpoint performance)

As Victor suggested to me in call the other day, it could be very ergonomic to have a custom actix extractor to obtain a connection without exposing the pool to the endpoints.

woshilapin commented on editoast/src/views/test_app.rs:
```diff
@@ -0,0 +1,169 @@
+where
+    S: Service<Request, Response = ServiceResponse<BoxBody>, Error = Error>,
+{
+    #[allow(dead_code)]
```
Why is this `dead_code`? The entire module is protected by `#[cfg(test)]` so I would expect all of these functions to be used in test code at least? If there is a real reason, maybe we need a comment explaining why we allow `dead_code` here.

woshilapin commented on editoast/src/views/test_app.rs:
```diff
@@ -0,0 +1,169 @@
+        self.core_client.clone()
+    }
+
+    #[allow(dead_code)]
```
Same than for the function `core_client` above.

Wadjetz replied:
No test "v2" use this for the moment, but we will use it soon

woshilapin commented on editoast/src/views/test_app.rs:
```diff
@@ -0,0 +1,169 @@
+impl TestAppBuilder {
+    pub fn new() -> Self {
+        Self {
+            db_pool: None,
+            core_client: None,
+            db_pool_v1: false,
```
This pattern looks like a `enum`.

```rust
enum DbPoolVersion {
  V1,
  V2(DbConnectionPoolV2),
}
```

This would avoid all the asserts below I believe.

woshilapin replied:
OK, after discussion, I realize that v1 is not here to stay and will disappear eventually. On top of that, the current structure is a builder, and having inconsistencies in a builder is ok since the `.build()` will check everything's fine before building the final object. And finally, it's going to be easier to remove v1 with the current `struct`, an `enum` would make that work a little harder. Let's keep it like this.

woshilapin commented on editoast/src/views/test_app.rs:
```diff
@@ -0,0 +1,169 @@
+    RedisClient,
+};
+
+pub(crate) struct TestAppBuilder {
```
It's not clear why we need a version 1 and a version 2. Could you put some comments explaining why?

woshilapin reviewed (commented):
It seems the complexity of the PR increased significantly since last time I reviewed it. I don't really understand why yet. I've made a bunch of comments, maybe not all make sense. We might need to discuss it.

Wadjetz committed:
editoast: fix documentation text

Co-authored-by: Jean SIMARD <woshilapin@tuziwo.info>
Update editoast/src/modelsv2/database/connection_pool.rs

Co-authored-by: Jean SIMARD <woshilapin@tuziwo.info>

fixup! editoast: add TestAppBuilder for DbConnectionPoolV2 and refactor setups

woshilapin reviewed (commented):
I resolved some of the comments and left the one I believe were either not clear, or not addressed. Tell me if you need any help, or if any comment is not understandable.

Wadjetz committed:
editoast: refactor functions organisation
